implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceFactory.java

```diff
@@ -0,0 +1,247 @@
+            final URL profileToFileName = toURL(addProfileToFileName(uri, profiles.get(i)));
+            consumeAsPath(profileToFileName, path -> addProfileConfigSource(profileToFileName, lastOrdinal + 1, configSources));
+        }
+    }
+
+    private static URL toURL(final URI uri) {
+        try {
+            return uri.toURL();
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    private void addConfigSource(final URI uri, final List<ConfigSource> configSources) {
+        addConfigSource(toURL(uri), configSources);
+    }
+
+    private void addConfigSource(final URL url, final List<ConfigSource> configSources) {
+        try {
+            final ConfigSource configSource = loadConfigSource(url);
+            if (!configSource.getProperties().isEmpty()) {
+                configSources.add(configSource);
+            }
+        } catch (IOException e) {
+            throw ConfigMessages.msg.failedToLoadResource(e);
+        }
+    }
+
+    private void addProfileConfigSource(final URL url, final int ordinal, final List<ConfigSource> configSources) {
+        try {
+            final ConfigSource configSource = loadConfigSource(url, ordinal);
+            if (!configSource.getProperties().isEmpty()) {
+                configSources.add(configSource);
+            }
+        } catch (FileNotFoundException e) {
+            // It is ok to not find the resource here, because it is an optional profile resource.
+        } catch (IOException e) {
+            throw ConfigMessages.msg.failedToLoadResource(e);
+        }
+    }
+
+    private boolean validExtension(final Path fileName) {
+        return validExtension(fileName.getFileName().toString());
+    }
+
+    private boolean validExtension(final String resourceName) {
+        for (String s : getFileExtensions()) {
+            if (resourceName.endsWith(s)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static URI addProfileToFileName(final URI uri, final String profile) {
+        final String fileName = uri.getPath() != null ? uri.getPath() : uri.getRawSchemeSpecificPart();
+        assert fileName != null;
+
+        final int dot = fileName.lastIndexOf(".");
+        final String fileNameProfile;
+        if (dot != -1) {
+            fileNameProfile = fileName.substring(0, dot) + "-" + profile + fileName.substring(dot);
+        } else {
+            fileNameProfile = fileName + "-" + profile;
+        }
+
+        return URI.create(uri.toString().replace(fileName, fileNameProfile));
+    }
+
+    private static class URIConverter implements Converter<URI> {
+        private static final long serialVersionUID = -4852082279190307320L;
+
+        @Override
+        public URI convert(final String value) {
+            try {
+                // TODO - validate path
+                return new URI(value);
```

Comment:
Just as an FYI, you can use `URI.create()` instead here, in which case instead of throwing `URISyntaxException` it will throw an `IllegalArgumentException`.  In fact, using this method, your converter *could* simply be `URI::create`.

Reply:
Sure, but since this `Converter` is applied to user input, we probably want to catch and throw the exception with our own custom message.